pkg/validate/security_context_linux.go

```diff
@@ -876,6 +879,11 @@ var _ = framework.KubeDescribe("Security Context", func() {
 			if !supportsUserNamespaces {
 				Skip("no runtime handler found which supports user namespaces")
 			}
+
+			pathIDMap := rootfsPath(resp.GetInfo())
+			if err := supportsIDMap(pathIDMap); err != nil {
```

Comment:
why don't we use the [runtime handler features](https://github.com/kubernetes-sigs/cri-tools/pull/1489/files#diff-009f9008c3866ca7eb564aa00ef9855c8c612b5c19b293909c56d44aac667779R870) feature? shouldn't containerd not report the feature is available if it's not?

Reply:
The runtime handler features reports if the OCI runtime knows about idmap mounts and how to handle them. The thing is, OCI spec mandates that unrecognized lines are ignored. So, an old OCI runtime that doesn't know about idmap mounts, will just ignore the lines "uidMapping" and so on the mounts.

The end result if that happens is horrible, because there is no mapping applied and then you use the volume with the hostUID (the ID on the host the container is mapped to) and then create files. The hostUID changes every time you create the pod, so the volume will be with garbage in the UID/GID of files.

This is why we added to the runtime-spec the way to report in features if idmap mounts are supported: https://github.com/opencontainers/runtime-spec/pull/1219

That is to avoid that situation: you just don't give idmap mounts to runtime that don't know about it. Because they will ignore it and create the aforementioned mess.

So once you know the runtime will handle it properly (using the features you mentioned), you can pass it to the runtime. But idmap mount support is filesystem-dependant. That means, each filesystem used needs to support it. And more filesystems are adding support every kernel release.

So you send the request to the OCI runtime to use idmap mounts on some volumes, but the runtime can return an error if the given filesystem doesn't support idmap mounts on the kernel it is running.

In particular in AlmaLinux 8 what is happening is: runc is new enough to handle them correctly, so we can pass a config.json with idmap mounts to that runc version. So we do. But the kernel is old and doesn't support the syscalls needed (so it doesn't support idmap mounts), so runc returns an error when trying to set it up.

What this PR does is just testing for idmap mounts supports and, if it is not supported, it skips the test.

tl; dr: the features command is useful to know if the OCI runtime knows about idmap mounts at compile time (see more info on the problem it solves here: https://github.com/opencontainers/runtime-spec/pull/1219). However, we can run runc in filesystems that don't support idmap mounts (one fs can support it but another don't), and runc will return an error in those cases. We just skip the tests instead, to avoid the expected error.

Reply:
Maybe the shorter answer is: containerd can only report it will be used or fail, but not ignored (as it was before if containerd/runc didn't know about idmap mounts and userns).

But each file-system used by mounts on the container needs to support idmap mounts. So it can't report yes/no without knowing which pod (e.g. if the pod users NFS, then it will fail, but if only uses ext4 and tmpfs, it will work on lot of kernels).

Reply:
what file system are we trying to create a mount here? if runc thinks alma support idmapped mounts I'd expect all the mounts we create to work. like, we control what environments we run these tests in. We could make sure they're not nfs but instead ext4 or tmpfs (whatever the alma kernel supports at that time). What file systems are these tests running on that are failing, but runc is saying it's supporteD?